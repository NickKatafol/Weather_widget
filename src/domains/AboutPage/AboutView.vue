<script setup lang="ts">
import {taskText} from "./assets/taskText"
import {useRouter, useRoute} from "vue-router"
import { RouterLink } from 'vue-router'

const router = useRouter()
const route = useRoute()

</script>


<template>
  <div class="about-wrapper">
    <header>
      <h1># ABOUT</h1>
      <div @click="router.push('/')" class="header__cross-btn"></div>
    </header>
  
    <article v-html="taskText">
    </article>
  
    <footer>
      <div @click="$router.push('/')"
           class="controller_open-space"
      >
        BACK to Home
      </div>
    </footer>
  </div>
</template>


<style scoped lang="scss">
@import "@/assets/CSS/utils.scss";

.about-wrapper {
  box-sizing: border-box;
  padding: calc(var(--vt-c-common-padding)*2);
  
  header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    
    .header__cross-btn {
      width: rem(30);
      height: rem(30);
      margin-left: auto;
      cursor: pointer;
      background-image: url("@/domains/AboutPage/assets/imgs/adminImgs/close.svg");
      background-size: rem(30) rem(30);
    }

    .header__cross-btn:hover {
      opacity: 0.2;
    }
  }

  article {
    white-space: pre-wrap;
  }
  
  footer {
    margin-top: calc(var(--vt-c-common-padding)*2);
  }
}

</style>
