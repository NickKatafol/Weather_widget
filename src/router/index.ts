import { createRouter, createWebHistory } from 'vue-router'
import Weather from '../domains/Weather/Weather.vue'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'weather',
      component: Weather
    },
    {
      path: '/about',
      name: 'about',
      component: () => import('../domains/AboutPage/AboutView.vue')
    }
  ]
})

export default router
