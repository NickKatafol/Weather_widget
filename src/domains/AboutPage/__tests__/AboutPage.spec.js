import {render, screen, fireEvent, cleanup, getByTestId} from "@testing-library/vue"
import '@testing-library/jest-dom'
import {waitFor} from '@testing-library/dom'  // но инстиллировать пакет НЕ требуется

import router from "../../../router"
import AboutView from '../AboutView.vue'

describe('Testing the Vue-components', () => {
  let wrapper

  beforeEach(() => {
    wrapper = render(
      AboutView,
      {
        global: {
          plugins: [router],
        },
      }
    )
  })

  afterEach(() => cleanup())


  it('Render Test - AboutView.vue is rendered properly', async () => {
    const headerEl = await screen.findByText("# ABOUT")

    expect(headerEl).toHaveTextContent("ABOUT")
    expect(wrapper).toMatchSnapshot()
  })


  it('Click & Router Test - Check the Router.push work out when cross-button click', async () => {
    //changing the initial value of router.currentRoute._value.path
    //which was '/'
    await router.push('/farAway')
    await fireEvent.click(wrapper.container.querySelector('.header__cross-btn'))

    await waitFor(() => {
        expect(router.currentRoute._value.path).toBe('/')
    })
  })


  it('Click & Router Test - Check the Router.push work out when click "BACK to Home" button on', async () => {
    const routerPush = vi.spyOn(router, 'push')

    await fireEvent.click(screen.getByText('BACK to Home'))

    expect(routerPush).toHaveBeenCalled()
    expect(routerPush).toHaveBeenCalledWith('/')
  })
})
