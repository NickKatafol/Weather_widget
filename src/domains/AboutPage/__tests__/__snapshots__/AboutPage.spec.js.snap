// Vitest Snapshot v1

exports[`Testing the Vue-components > Render Test - AboutView.vue is rendered properly 1`] = `
{
  "baseElement": <body>
    <div>
      <div
        class="about-wrapper"
        data-v-3c7b92b3=""
      >
        <header
          data-v-3c7b92b3=""
        >
          <h1
            data-v-3c7b92b3=""
          >
            # ABOUT
          </h1>
          <div
            class="header__cross-btn"
            data-v-3c7b92b3=""
          />
        </header>
        <article
          data-v-3c7b92b3=""
        >
          

          <b>
            Frontend developer demo
          </b>
          

          <br />
          
Implement a widget that displays the weather.

          <br />
          

          <br />
          

          <span
            style="color: var(--vt-c-fiolet);"
          >
            # Technologies:
          </span>
          

          <br />
          Vue.js - 3, Pinia, Vuelidate, ohmyfetch, 

          <br />
          TypeScript, class, architects skills - DDD, SOLID, TDD, minor dev rules, 

          <br />
          ES7+, 

          <br />
          SCSS, grid, flexbox,

          <br />
          Vitest (unit testing), Cypress (e2e testing).

          <br />
          

          <br />
          

          <span
            style="color: var(--vt-c-fiolet);"
          >
            # Conditions:
          </span>
          

          <br />
          
1. Data for the widget taken from free public API: 

          <br />
          
https://openweathermap.org/current,

          <br />
          via ohmyfetch through 
          <b>
            apiService module
          </b>
          .

          <br />
          

          <br />
          
2. 
          <b>
            API errors
          </b>
           is handlered with a message:

          <br />
          - entering wrong city name while pointing weather location

          <br />
          

          <br />
          
3. After clicking the ‘Gear’ button at the upper-right corner, the view switch to the following:

          <br />
          
Here, a user be able to:

          <br />
          
a. Remove previously added location.

          <br />
          
b. Reorder the location by 
          <b>
            dragging and dropping
          </b>
           them within the list.

          <br />
          
c. Add new locations.
There is a 
          <b>
            validation
          </b>
           with number within certain interval allowing negative numbers.

          <br />
          

          <br />
          
4. Save the configuration in the 
          <b>
            local storage
          </b>
           and restore it on future visits. A user is be able
to configure the widget just once and then have the same view until they change the computer
or clean the storage.

          <br />
          

          <br />
          
5. If the local storage is empty it take current 
          <b>
            geolocation
          </b>
           coordinates.

          <br />
          

          <br />
          
6. Use the 
          <b>
            universal applicable vue component
          </b>
           - '@/components/gridCard/GridCard.vue'.

          <br />
          

          <br />
          
7. Make a bunch of tests where 
          <b>
            unit
          </b>
           and 
          <b>
            e2e
          </b>
           tests cover code with 
          <b>
            Vitest
          </b>
           and 
          <b>
            Cypress.io
          </b>
          

          <br />
          Unit tests include:

          <br />
          - Render component with props, child components, mocking Pinia,

          <br />
          - Click & Router Test - Check the Router.push work out when cross-button click,

          <br />
          - Exported function Testing, test is inside of JS file,

          <br />
          - Pinia Testing - testing actions, setters and getters,

          <br />
          - Testing static methods of class

          <br />
          - as examples it is probably enough.

          <br />
          

          <br />
          E2E tests include:

          <br />
          - routing between pages

          <br />
          - examine a bunch of business processes

          <br />
          

          <br />
          

          <span
            style="color: var(--vt-c-fiolet);"
          >
            # Notes:
          </span>
          

          <br />
          KEEP IN MIND

          <br />
          There is not the production issue. 

          <br />
          First at all it's just demonstration the developer style end some skills.

          <br />
          It is possible to expand and deepen the presented backbone in a multifaceted and elaborate way.

          <br />
          

          <br />
          For example there is no objective to cover code with tests tremendously. I only point to my useful skills about it.

        </article>
        <footer
          data-v-3c7b92b3=""
        >
          <div
            class="controller_open-space"
            data-v-3c7b92b3=""
          >
             BACK to Home 
          </div>
        </footer>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="about-wrapper"
      data-v-3c7b92b3=""
    >
      <header
        data-v-3c7b92b3=""
      >
        <h1
          data-v-3c7b92b3=""
        >
          # ABOUT
        </h1>
        <div
          class="header__cross-btn"
          data-v-3c7b92b3=""
        />
      </header>
      <article
        data-v-3c7b92b3=""
      >
        

        <b>
          Frontend developer demo
        </b>
        

        <br />
        
Implement a widget that displays the weather.

        <br />
        

        <br />
        

        <span
          style="color: var(--vt-c-fiolet);"
        >
          # Technologies:
        </span>
        

        <br />
        Vue.js - 3, Pinia, Vuelidate, ohmyfetch, 

        <br />
        TypeScript, class, architects skills - DDD, SOLID, TDD, minor dev rules, 

        <br />
        ES7+, 

        <br />
        SCSS, grid, flexbox,

        <br />
        Vitest (unit testing), Cypress (e2e testing).

        <br />
        

        <br />
        

        <span
          style="color: var(--vt-c-fiolet);"
        >
          # Conditions:
        </span>
        

        <br />
        
1. Data for the widget taken from free public API: 

        <br />
        
https://openweathermap.org/current,

        <br />
        via ohmyfetch through 
        <b>
          apiService module
        </b>
        .

        <br />
        

        <br />
        
2. 
        <b>
          API errors
        </b>
         is handlered with a message:

        <br />
        - entering wrong city name while pointing weather location

        <br />
        

        <br />
        
3. After clicking the ‘Gear’ button at the upper-right corner, the view switch to the following:

        <br />
        
Here, a user be able to:

        <br />
        
a. Remove previously added location.

        <br />
        
b. Reorder the location by 
        <b>
          dragging and dropping
        </b>
         them within the list.

        <br />
        
c. Add new locations.
There is a 
        <b>
          validation
        </b>
         with number within certain interval allowing negative numbers.

        <br />
        

        <br />
        
4. Save the configuration in the 
        <b>
          local storage
        </b>
         and restore it on future visits. A user is be able
to configure the widget just once and then have the same view until they change the computer
or clean the storage.

        <br />
        

        <br />
        
5. If the local storage is empty it take current 
        <b>
          geolocation
        </b>
         coordinates.

        <br />
        

        <br />
        
6. Use the 
        <b>
          universal applicable vue component
        </b>
         - '@/components/gridCard/GridCard.vue'.

        <br />
        

        <br />
        
7. Make a bunch of tests where 
        <b>
          unit
        </b>
         and 
        <b>
          e2e
        </b>
         tests cover code with 
        <b>
          Vitest
        </b>
         and 
        <b>
          Cypress.io
        </b>
        

        <br />
        Unit tests include:

        <br />
        - Render component with props, child components, mocking Pinia,

        <br />
        - Click & Router Test - Check the Router.push work out when cross-button click,

        <br />
        - Exported function Testing, test is inside of JS file,

        <br />
        - Pinia Testing - testing actions, setters and getters,

        <br />
        - Testing static methods of class

        <br />
        - as examples it is probably enough.

        <br />
        

        <br />
        E2E tests include:

        <br />
        - routing between pages

        <br />
        - examine a bunch of business processes

        <br />
        

        <br />
        

        <span
          style="color: var(--vt-c-fiolet);"
        >
          # Notes:
        </span>
        

        <br />
        KEEP IN MIND

        <br />
        There is not the production issue. 

        <br />
        First at all it's just demonstration the developer style end some skills.

        <br />
        It is possible to expand and deepen the presented backbone in a multifaceted and elaborate way.

        <br />
        

        <br />
        For example there is no objective to cover code with tests tremendously. I only point to my useful skills about it.

      </article>
      <footer
        data-v-3c7b92b3=""
      >
        <div
          class="controller_open-space"
          data-v-3c7b92b3=""
        >
           BACK to Home 
        </div>
      </footer>
    </div>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "html": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
